### PR TITLE
Support Rack v3

### PR DESCRIPTION
**Pull request by jeremiahlukus**
Rack v3 removed[ Rack::VERSION](https://github.com/rack/rack/blob/main/CHANGELOG.md#removed-1) and replaced it with Rack.release 

Master has Rack::VERSION [added back ](https://github.com/rack/rack/blob/main/CHANGELOG.md#added) so this will probably be permanent. 

Rack v3 [lowercases all headers](https://github.com/rack/rack/blob/main/UPGRADE-GUIDE.md#rack-2--rack-3-compatibility) `Response header keys can no longer include uppercase characters.`  So i added support for lowercase headers.

The key points of supporting Rack v3 is the response array must be mutable, the status is an integer, headers are a mutable hash with lowercase keys.

### CODE REVIEW

**Comment by jeremiahlukus** [comment]
I want to test a bit more today and over the next few days before merging

**Comment by jeremiahlukus** [comment]
@metaskills  I want to run run tests against both rackv2 and rackv3 using 
https://github.com/thoughtbot/appraisal 

I think this will help since all tests pass using rackv2 however using rackv3 tests fail. Having both will hopefully avoid some issues. 

Do you have any issues with that? 
I want to separate tests into to dir's [testRackv2](https://github.com/rails-lambda/lamby/pull/181/files#diff-ee98e028c59b193d58fde56ab4daf54d43c486ae674e63d50ddf300b07943e0fR6) and testRackv3. Then in the [bin/test ](https://github.com/rails-lambda/lamby/pull/181/files#diff-4e821476d5e4f5626c6f9fca337a47afe96a5e3daf23bccb02ad0bb109a9a407R11) run both (v2 is commented out since they pass v3 does not pass so working on that)

**Comment by jeremiahlukus** [comment]
Screenshot 2024-07-12 at 5 03 06 PM


All tests are passing locally now. 


@metaskills  to give you an overview of the changes 

In bin/setup i added `bundle exec appraisal install` which bundle installs rails 7.0 with rack 2.2.8 this also bundle installed rails 7.1 which uses rack 3.1.7

In bin/test Im running tests again rack v2 and then after those are finished rack v3. 


in the test/ dir i created two additional folders testRackv2 and testRackv3 these just have the hander test and proxy server test. I had to copy over the dummy app and test helpers as well since they are separate.   



Since there are opinions littered in the PR would like to get your say haha. I do strongly want to test over rack v2 and rack v3 though. I am not attached to the way i went about doing it but its the cleanest way i could think of.

File: .DS_Store
Comment (metaskills) [inline .DS_Store:1]: Recommend setting up a global .gitignore file and adding .DS_Store to it.

**Review by metaskills — CHANGES_REQUESTED** [review]
Given that we have to contextualize the entire test directory to v2 and v3 for rack, my question is should we just cut Lamby v6 and say that is rack v3 and up? So basically we can do like Rails and create a 5-stable branch to take security patches for 5.x (rack v2) if needed and then move main (v6 with rack v3) compatibility forward. Obvious, if we did this, main would have to set a gem spec change of rack v3 or higher. Does that make sense? This would also help avoid the dupdated `test/testRackv3` directories which the names look odd to me. We do not use came case in any other dir names.

**Comment by jeremiahlukus** [comment]
`which the names look odd to me.` Yea my bad i pushed to snake cased them and make them have a better name 


`my question is should we just cut Lamby v6 and say that is rack v3 and up?` For me it does feel right to cut a v6 and drop rack v2 since it works fine. The only thing that is different is the tests. 

`we can do like Rails and create a 5-stable branch` again its just hard to justify adding that since it works with both rackv2 and v3. 


The two tests that are different it the hander_test and the proxy_server_test the other stuff on those dir's are just helper files. I cleaned up the code a bit so now it might look a bit better to you. Now in `test/rackv2` and `test/rackv3` there are only 3 files  `handler_test, proxy_server_test and test_helper`

**Comment by metaskills** [comment]
> For me it does feel right to cut a v6 and drop rack v2 since it works fine. The only thing that is different is the tests.

Yup, it seems odd but I like it. A few reasons why. 1) if all the tests change, then I think that makes the case 2) It is not like we are doing massive development here. 3) Lamby is essentially a rack web server. It is listed on Rack as such.

> again its just hard to justify adding that since it works with both rackv2 and v3.

Making the branch is easy. The idea is how much work are we going to do on that branch afterward. I'm guessing not that much. I like following Rails here since it too is a Rack app. IE, Pick a version where v3 is supported. Declare that as a hard dep. Move forward with good backward compatibility. Oh... that means 5-stable branch would need a v2 lock down in the gemspece and version push too.

**Comment by jeremiahlukus** [comment]
Ok sounds good, do you mind doing the v5-stable branch and I’ll fix this PR and add the rack 3 lock. Hopefully tomorrow 👍.

**Comment by metaskills** [comment]
Yup, just pushed it up and created 5.2.2. How does it look to you? If good, I can push that to Rubygems.

**Comment by jeremiahlukus** [comment]
Looks good to me, send it!

**Comment by jeremiahlukus** [comment]
@metaskills  Ok just pushed to set the version to 6.0.0, drop support for rack v2, and add support for rack v3.

**Comment by metaskills** [comment]
> Looks good to me, send it!

Ok, I just pushed v5.2.2 from the 5-stable branch.

**Review by metaskills — APPROVED** [review]
Have not done a deep dive but I think everything is in order. @jeremiahlukus I really want to thank you so much for contributing here and bringing new live into the project. I owe you big time. If there is anything I can do let me know.

**Comment by jeremiahlukus** [comment]
Sweet, most of the changes are from tests. 

The code is just to_s body/headers, use add lowercase headers and use either rack version or rack release.

**Pull request merged**